r2/service.go

```diff
@@ -59,86 +64,480 @@ func NewService(iOpts instrument.Options, rootPrefix string, store Store) server
 	return &service{iOpts: iOpts, store: store, rootPrefix: rootPrefix}
 }
 
+type payload struct {
+	data   []byte
+	status int
+}
+
+func newPayload(d []byte, status int) payload {
+	return payload{data: d, status: status}
+}
+
+func (p payload) write(w http.ResponseWriter) {
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(p.status)
+	w.Write(p.data)
+}
+
+func writeAPIResponse(w http.ResponseWriter, code int, msg string) {
+	j, _ := json.Marshal(apiResponse{Code: code, Message: msg})
+	newPayload(j, code).write(w)
+}
+
+func handleError(w http.ResponseWriter, opError error) {
+	switch opError.(type) {
+	case ConflictError:
+		writeAPIResponse(w, http.StatusConflict, opError.Error())
+	case BadInputError:
+		writeAPIResponse(w, http.StatusBadRequest, opError.Error())
+	case VersionError:
+		writeAPIResponse(w, http.StatusConflict, opError.Error())
+	case NotFoundError:
+		writeAPIResponse(w, http.StatusNotFound, opError.Error())
+	default:
+		writeAPIResponse(w, http.StatusInternalServerError, opError.Error())
+	}
+}
+
+// TODO(dgromov): Make this return a list of validation errors
+func parseRequest(s interface{}, body io.ReadCloser) string {
+	if err := json.NewDecoder(body).Decode(s); err != nil {
+		return fmt.Sprintf("Malformed Json: %s", err.Error())
+	}
+
+	validate := validator.New()
+
+	validate.RegisterTagNameFunc(func(fld reflect.StructField) string {
+		name := strings.SplitN(fld.Tag.Get("json"), ",", 2)[0]
+		return name
+	})
+
+	required := make([]string, 0)
+
+	if err := validate.Struct(s); err != nil {
+		for _, e := range err.(validator.ValidationErrors) {
+			if e.ActualTag() == "required" {
+				required = append(required, e.Namespace())
+			}
+		}
+	}
+
+	if len(required) > 0 {
+		return fmt.Sprintf("Required: [%v]", strings.Join(required, ", "))
+	}
+	return ""
+}
+
 func (s *service) readNamespaces(w http.ResponseWriter, _ *http.Request) {
-	w.WriteHeader(http.StatusNotImplemented)
-	fmt.Fprintf(w, "Placeholder for readNamespace")
+	names, v, err := s.store.FetchNamespaces()
+	if err != nil {
+		writeAPIResponse(w, http.StatusInternalServerError, "Could not read namespaces")
+		return
+	}
+	nss := make([]namespaceJSON, len(names))
+	for i, n := range names {
+		nss[i] = namespaceJSON{ID: n}
+	}
+
+	j, err := json.Marshal(namespacesJSON{
+		Version:    v,
+		Namespaces: nss,
+	})
+
+	if err != nil {
+		writeAPIResponse(w, http.StatusInternalServerError, "Could not create response object")
```

Comment:
same as above, log if writeAPIResponse fails. Need to do this in all function below too